src/label.rs

```diff
@@ -366,8 +371,42 @@ pub struct Label {
     pub polarity: Polarity,
     /// The path of the type being currently checked in the original type.
     pub path: ty_path::Path,
+    /// An environment mapping type variables to [`TypeVarData`]. Used by
+    /// polymorphic contracts to decide which actions to take when encountering a `forall`.
+    pub type_environment: HashMap<SealingKey, TypeVarData>,
 }
 
+/// Data about type variables that is needed for polymorphic contracts to decide which actions to take.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TypeVarData {
+    pub polarity: Polarity,
```

Comment:
I guess this field will take on more values once you do the "one-pass" version? Morally, `positive`, `negative`, and the composite partial identities `(positive, negative)` and `(negative, positive)` ?

Reply:
I was expecting to need more data for type variables, indeed. Originally I was also tracking the variable kind, but that ended up being unnecessary for now.